src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -10797,44 +10881,13 @@ namespace WaterThermalTanks {
                 }
 
                 // next determine if tank temperature is such that source side flow might be requested
-                if (!WaterThermalTank(WaterThermalTankNum).IsChilledWaterTank) {
-                    if (OutletTemp < DeadBandTemp) {
-                        NeedsHeat = true;
-                    } else if ((OutletTemp >= DeadBandTemp) && (OutletTemp < SetPointTemp)) {
-                        // inside the deadband, use saved mode from water heater calcs
-                        if (WaterThermalTank(WaterThermalTankNum).SavedMode == HeatMode) {
-                            NeedsHeat = true;
-                        } else if (WaterThermalTank(WaterThermalTankNum).SavedMode == FloatMode) {
-                            NeedsHeat = false;
-                        }
-                    } else if (OutletTemp >= SetPointTemp) {
-                        NeedsHeat = false;
-                    }
-                } else { // is a chilled water tank so flip logic
-                    if (OutletTemp > DeadBandTemp) {
-                        NeedsCool = true;
-                    } else if ((OutletTemp <= DeadBandTemp) && (OutletTemp > SetPointTemp)) {
-                        // inside the deadband, use saved mode from water thermal tank calcs (modes only for mixed)
-                        if (WaterThermalTank(WaterThermalTankNum).TypeNum == MixedChilledWaterStorage) {
-                            if (WaterThermalTank(WaterThermalTankNum).SavedMode == CoolMode) {
-                                NeedsCool = true;
-                            } else if (WaterThermalTank(WaterThermalTankNum).SavedMode == FloatMode) {
-                                NeedsCool = false;
-                            }
-                        } else if (WaterThermalTank(WaterThermalTankNum).TypeNum == StratifiedChilledWaterStorage) {
-                            NeedsCool = true;
-                        }
-
-                    } else if (OutletTemp <= SetPointTemp) {
-                        NeedsCool = false;
-                    }
-                }
+                NeedsHeatOrCool = SourceHeatNeed(WaterThermalTankNum,OutletTemp, DeadBandTemp,SetPointTemp);
```

Comment:
I very much like this duplicated code is extracted out into a worker function.  The format is messy here though, need to include spaces after each comma in an argument list.